### PR TITLE
add tests for issues #10207, #10178, #12939

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

File: test/core.jl [hunk 1]
```diff
@@ -81,6 +81,22 @@ _bound_vararg_specificity_1{T}(::Type{Array{T,1}}, d::Int) = 1
 @test _bound_vararg_specificity_1(Array{Int,1}, 1) == 1
 @test _bound_vararg_specificity_1(Array{Int,2}, 1, 1) == 0
 
+# issue #12939
+module Issue12939
+abstract type Abs; end
+struct Foo <: Abs; end
+struct Bar; val::Int64; end
+struct Baz; val::Int64; end
+f{T}(::Type{T}, x::T) = T(3)
```
Comment (ararslan) [inline test/core.jl]: Any reason not to use `where` here?

**Comment by ararslan** [comment]
Can the linked issues be closed once this PR is merged? If so, you could add text to the commit message that would auto-close them.

File: test/core.jl (hunk 1 above)
Comment (JeffBezanson) [inline test/core.jl]: I was just copying this from the issue comment.

**Comment by JeffBezanson** [comment]
Yes, the linked issues are fixed now. Easier to just close them than to update the commit I think.

**Pull request merged**